File: leetcode/balanced_binary_tree.js
/* Given a binary tree, determine if it is height-balanced.
 * For this problem, a height-balanced binary tree is defined as a binary tree 
 * in which the depth of the two subtrees of every node never differ by more 
 * than 1.
 * Time Complexity: O(n)
 * Space Complexity: O(h), h is the height of the tree
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
    function helper(root) {
        if (!root)
            return 0;
        var left_height = helper(root.left);
        if (left_height < 0)
            return -2;
        var right_height = helper(root.right);
        if (Math.abs(left_height - right_height) > 1) 
            return -2;
        return 1 + Math.max(left_height, right_height);
    }
    return helper(root) >= 0;
};

File: lintcode/balanced_binary_tree.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    """
    @param root: The root of binary tree.
    @return: True if this Binary tree is Balanced, or false.
    """
    def isBalanced(self, root):
        def helper(root):
            if not root: return 0
            left_height = helper(root.left)
            if left_height < 0: return -2
            right_height = helper(root.right)
            if abs(left_height - right_height) > 1: return -2
            return 1 + max(left_height, right_height)
        return helper(root) >= 0
